compiler/damlc/daml-compiler/src/DA/Daml/Compiler/DataDependencies.hs

```diff
@@ -163,16 +166,67 @@ generateSrcFromLf env = noLoc mod
             lname = mkRdrName (LF.unExprValName lfName) :: Located RdrName
             sig = TypeSig noExt [lname] . HsWC noExt . HsIB noExt <$> ltype
             lsigD = noLoc . SigD noExt <$> sig :: Gen (LHsDecl GhcPs)
-            lexpr = noLoc $ HsPar noExt $ noLoc $ HsVar noExt lname :: LHsExpr GhcPs
-            lgrhs = noLoc $ GRHS noExt [] lexpr :: LGRHS GhcPs (LHsExpr GhcPs)
-            grhss = GRHSs noExt [lgrhs] (noLoc $ EmptyLocalBinds noExt)
-            matchContext = FunRhs lname Prefix NoSrcStrict
-            lmatch = noLoc $ Match noExt matchContext [] Nothing grhss
-            lalts = noLoc [lmatch]
-            bind = FunBind noExt lname (MG noExt lalts Generated) WpHole []
+            bind = mkTrivialBind lname
             lvalD = noLoc $ ValD noExt bind :: LHsDecl GhcPs
         [ lsigD, pure lvalD ]
 
+    -- | Generate instance declarations from dictionary functions.
+    instanceDecls :: [Gen (LHsDecl GhcPs)]
+    instanceDecls = do
+        LF.DefValue {..} <- NM.toList $ LF.moduleValues $ envMod env
+        let dvalType = snd dvalBinder
+        Just qname@LF.Qualified{..} <- [getDFunClassName dvalType]
+        Just fieldNames <- [getDFunFieldNames dvalBody]
+        guard (not (isHasField qname))
+        pure $ do
+            polyTy <- HsIB noExt . noLoc <$> convType env dvalType
+            ghcMod <- genModule env qualPackage qualModule
+            pure . noLoc . InstD noExt . ClsInstD noExt $ ClsInstDecl
+                { cid_ext = noExt
+                , cid_poly_ty = polyTy
+                , cid_binds = listToBag (mapMaybe (mkBind ghcMod) fieldNames)
+                , cid_sigs = []
+                , cid_tyfam_insts = []
+                , cid_datafam_insts = []
+                , cid_overlap_mode = Nothing
+                }
+
+      where
+
+        -- | Filter out HasField instances, since they are generated separately.
+        isHasField :: LF.Qualified LF.TypeSynName -> Bool
```

Comment:
Where are they generated? I thought we don’t run the preprocessor on generated sources. Might be good to add a test for this (if it doesn’t work, let’s do it in a separate PR).

Reconstructing those instances from DAML-LF is a bit tricky since the type-level strings are lost. For `HasField` we can probably get the field name fairly easily from the body. A bit gross but probably fairly reasonable. The only other instance that I remember that relies heavily on type-level stuff that is lost in the translation to LF is `Generic` and for now, I would just ignore that (but probably also good to have a testcase for this, happy to do it in a separate PR)

Reply:
We could just run the daml-preprocessor step for records, rather than recreating that. Since we'll (soon) re-export the typeclasses, we don't even need to change the prefix.

(We should also be able to drop the `CurrentSdk` logic from the enum type preprocessor as well.)

I'll handle this in a separate PR.

Reply:
The problem with that is that then you will get the stub instances even if the DALF does not contain the instances. Probably not a huge deal in practice but it feels a bit fragile.

Reply:
Hmm, I see. I guess the safer thing is to make the stub instance by reconstructing the field name from the body of the HasField instance. That's definitely a separate PR.